### PR TITLE
🏗 Fix dependency extraction while watching JS entry points

### PR DESCRIPTION
**Pull request by rsimha**
#33840 introduced a way to extract the dependencies of JS entry points so that the watcher could intelligently react to code changes. This broke in situations when an extension contained JSX, because `getDependencies()` wasn't using all the plugins required to parse and transform the source code.

image

This PR makes `getDependencies()` use the exact set of plugins that would've been used by `esbuild` during normal compilation. With this, watching extensions containing JSX code is fixed, and future changes to the set of babel plugins used during compilation will automatically get picked up by `getDependencies()`.

image

Addresses https://github.com/ampproject/amphtml/pull/33873#issuecomment-821327849
Implements https://github.com/ampproject/amphtml/pull/33873#issuecomment-821334930

### CODE REVIEW

File: build-system/tasks/helpers.js [hunk 1]
```diff
@@ -827,31 +827,23 @@ function mkdirSync(path) {
...
- * Returns the list of dependencies for a given JS entrypoint.
+ * Returns the list of dependencies for a given JS entrypoint by having esbuild
+ * generate a metafile for it. Uses the babel plugin that would've been used to
+ * build the entrypoint.
  * @param {string} entryPoint
+ * @param {!Object} options
  * @return {Promise<Array<string>>}
  */
-async function getDependencies(entryPoint) {
-  const stripImportAssertionPlugin = {
```
Comment (alanorozco) [inline build-system/tasks/helpers.js:835]: Why were we only doing this previously?

Comment (rsimha) [reply build-system/tasks/helpers.js:835]: I'm not sure. From my reading of the code, we should be running all plugins that would've been run during compilation, and only then extract the list of dependencies from `metafile`. This will ensure that if we add new plugins in future to handle other kinds of special code, `getDependencies()` will automatically use them.

I added some debugging statements and verified that this works across a selection of JS entry points and extensions.

Comment (rsimha) [reply build-system/tasks/helpers.js:835]: The original PR description https://github.com/ampproject/amphtml/pull/33840#issue-615533919 says the following:

> Creates a new esbuild plugin for stripping import assertions. This will become unecessary once import assertions become stage 4 (and are built-in to esbuild)

Given that the watcher is broken right now, I think we should fix it by using whatever babel transforms we'd have used during compilation, and if required in future, add an additional plugin that does assertion stripping. We should run this by @samouri when he's back next week, but for now, the approach in this PR fixes compilation for JSX extensions.

@alanorozco Does this SGTY?

Comment (alanorozco) [reply build-system/tasks/helpers.js:835]: Yep, thanks!

**Review by alanorozco — APPROVED** [review]

**Review by dmanek — APPROVED** [review]

File: build-system/tasks/helpers.js [hunk 2]
```diff
@@ -827,31 +826,24 @@ function mkdirSync(path) {
...
-        const contents = (await batchedRead(filename)).contents.toString();
-        const stripped = contents.replace(/assert {type: 'json'}/g, '');
-        return {contents: stripped};
-      });
-    },
-  };
-
+async function getDependencies(entryPoint, options) {
+  const caller = options.minify ? 'minified' : 'unminified';
+  const babelPlugin = getEsbuildBabelPlugin(caller, /* enableCache */ true);
```
Comment (jridgewell) [inline build-system/tasks/helpers.js:839]: Jake purposefully didn't use Babel here, because it's significantly slower. I think all that was necessary was to use the JSX esbuild loader, right?

Comment (rsimha) [reply build-system/tasks/helpers.js:839]: Here's my rationale behind using Babel in this PR:
- `getDependencies()` is called only while initially watching a main entry point, which are significantly fewer in number than JS files in total
- It doesn't really matter if the call to Babel is slightly slower because `getDependencies()` is only called once during startup
- The fact that caching is enabled for `getEsbuildBabelPlugin()` means that the subsequent transforms of the entry point will be instant (so this shouldn't actually slow down anything)
- This PR will make us immune to future breakages when a new kind of file is introduced (similar to JSX)

Wanna pass this through your BS detector and tell me if there's any reason not to leave things as they are? 😃

Comment (jridgewell) [reply build-system/tasks/helpers.js:839]: Sounds good to me.

File: build-system/tasks/helpers.js (hunk 1 above)
Comment (samouri) [reply build-system/tasks/helpers.js:835]: Initially I was hoping to exclude babel from transpilations for performance reasons. I had assumed `esbuild` had jsx transforms built-in (which I still think it does, I just didn't enable them). 

Import assertions aren't builtin yet, so (but are handled by one of our babel transforms) so I created a simple string replace to handle it. @rsimha's logic in a separate thread that the TransformCache means there is a negligible cost of using our full babel pipeline is a great point that I hadn't considered!

tldr: This all LGTM

File: build-system/tasks/helpers.js (hunk 2 above)
Comment (samouri) [reply build-system/tasks/helpers.js:839]: Running through my BS Detector:

> getDependencies() is called only while initially watching a main entry point, which are significantly fewer in number than JS files in total

This one is likely BS, since we actually perform a full `esbuild.build` operation here. Meaning it still transforms every file references by the entrypoint.

> getDependencies() is called only while initially watching a main entry point, which are significantly fewer in number than JS files in total

It'll be called once _per entrypoint_ (regardless of main vs. extension) during watch mode.

> The fact that caching is enabled for getEsbuildBabelPlugin() means that the subsequent transforms of the entry point will be instant (so this shouldn't actually slow down anything)

This is a fantastic point. I forgot that it becomes essentially free when cached (which we can rely on working during local dev / watch mode). Couple with the final point, I'm sold :)
